src/libcore/task/wake.rs

```diff
@@ -63,21 +67,122 @@ pub struct RawWakerVTable {
     /// pointer.
-    pub wake: unsafe fn(*const ()),
+    wake_by_ref: unsafe fn(*const ()),
 
     /// This function gets called when a [`RawWaker`] gets dropped.
     ///
     /// The implementation of this function must make sure to release any
     /// resources that are associated with this instance of a [`RawWaker`] and
     /// associated task.
-    pub drop: unsafe fn(*const ()),
+    drop: unsafe fn(*const ()),
+}
+
+impl RawWakerVTable {
+    /// Creates a new `RawWakerVTable` from the provided `clone`, `wake`, and
+    /// `drop` functions.
+    ///
+    /// # `clone`
+    ///
+    /// This function will be called when the [`RawWaker`] gets cloned, e.g. when
+    /// the [`Waker`] in which the [`RawWaker`] is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a [`RawWaker`] and associated
+    /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
+    /// of the same task that would have been awoken by the original [`RawWaker`].
+    ///
+    /// # `wake`
+    ///
+    /// This function will be called when `wake` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    ///
+    /// # `wake_by_ref`
+    ///
+    /// This function will be called when `wake_by_ref` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// This function is similar to `wake`, but must not consume the provided data
+    /// pointer.
+    ///
+    /// # `drop`
+    ///
+    /// This function gets called when a [`RawWaker`] gets dropped.
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    #[rustc_promotable]
+    #[unstable(feature = "futures_api",
+            reason = "futures in libcore are unstable",
+            issue = "50547")]
+    pub const fn new(
+        clone: unsafe fn(*const ()) -> RawWaker,
+        wake: unsafe fn(*const ()),
+        wake_by_ref: unsafe fn(*const ()),
+        drop: unsafe fn(*const ()),
+    ) -> Self {
+        Self {
+            clone,
+            wake,
+            wake_by_ref,
+            drop,
+        }
+    }
+}
+
+/// The `Context` of an asynchronous task.
+///
+/// Currently, `Context` only serves to provide access to a `&Waker`
+/// which can be used to wake the current task.
+pub struct Context<'a> {
+    waker: &'a Waker,
+    // Ensure we future-proof against variance changes by including
+    // an `&mut` reference with the provided lifetime.
+    _marker: PhantomData<&'a mut ()>,
```

Comment:
`&'a mut ()` is covariant in `'a`. This should be `fn(&'a ()) -> &'a ()`

Reply:
Discussed in previous PRs why it doesn't matter-- the primary concern is just making sure that if/when we add a mutable reference, things don't *change*. The particular variance of `&'a mut` doesn't matter so long as we're forward compatible with it. I don't think we're interested in being forwards-compatible with the addition of `fn(&'a T) -> &'a T` (that is to say, I don't think we *want* this lifetime to be invariant).

Reply:
Can you bake the discussion into the comment?

Reply:
Sure. (note that this change is part of in the previous PR in this set, so this will be added there and bubbled through)

Reply:
Meh, I guess it doesn't *hurt* to have it be invariant (I don't think there's anything that needs it to vary) so I guess we can give it a shot? it's backwards-compatible to relax it later.

Reply:
Done.